lib/Core/SQLiteBuildDB.cpp

```diff
@@ -285,22 +282,38 @@ class SQLiteBuildDB : public BuildDB {
   }
 
   void close() {
-    std::lock_guard<std::mutex> guard(dbMutex);
-    assert(db);
+    if (!db) return;
 
     // Destroy prepared statements.
     sqlite3_finalize(findKeyIDForKeyStmt);
+    findKeyIDForKeyStmt = nullptr;
     sqlite3_finalize(findKeyNameForKeyIDStmt);
+    findKeyNameForKeyIDStmt = nullptr;
     sqlite3_finalize(findRuleResultStmt);
+    findRuleResultStmt = nullptr;
     sqlite3_finalize(fastFindRuleResultStmt);
+    fastFindRuleResultStmt = nullptr;
     sqlite3_finalize(deleteFromKeysStmt);
+    deleteFromKeysStmt = nullptr;
     sqlite3_finalize(insertIntoKeysStmt);
+    insertIntoKeysStmt = nullptr;
     sqlite3_finalize(insertIntoRuleResultsStmt);
+    insertIntoRuleResultsStmt = nullptr;
 
     sqlite3_close(db);
     db = nullptr;
   }
 
+public:
+  SQLiteBuildDB(StringRef path, uint32_t clientSchemaVersion)
+    : path(path), clientSchemaVersion(clientSchemaVersion) { }
+
+  virtual ~SQLiteBuildDB() {
+    std::lock_guard<std::mutex> guard(dbMutex);
```

Comment:
This is a case where I think you actually don't need the lock -- if the db is being destructed while something else is operating on it that is already a bug?